File: PyProject/leetcode/history/quick_sort.py
def partition(arr, low, high):
    # 标记指针
    i = (low - 1)
    # 参考数
    pivot = arr[high]

    for j in range(low, high):
        if arr[j] < pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]

    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1


def quick_sort(arr, low, high):
    if low < high:
        pi = partition(arr, low, high)
        quick_sort(arr, low, pi - 1)
        quick_sort(arr, pi + 1, high)


if __name__ == '__main__':
    a = [0, 1, 199, 19, 10, 3, 8, 5, 2]
    quick_sort(a, 0, len(a) - 1)
    print(a)

File: GoProject/leetcode/sort/history/quick_sort.go
package main

func partition(nums []int, start int, end int) int {
	pivot := nums[end] // 基准值
	i := start - 1     // 最小元素索引

	for j := start; j < end; j++ {
		if nums[j] <= pivot {
			i += 1
			nums[i], nums[j] = nums[j], nums[i]
		}
	}
	nums[i+1], nums[end] = nums[end], nums[i+1]
	return i + 1
}

func quickSort(nums []int, start int, end int) {
	if start < end {
		pivot := partition(nums, start, end)
		quickSort(nums, start, pivot-1)
		quickSort(nums, pivot+1, end)
	}

}

func main() {
	arr := []int{10, 7, 8, 9, 1, 5}
	quickSort(arr, 0, len(arr)-1)

	for _, v := range arr {
		println(v)
	}
}
